File: ImgProc/tutorial_threshold/threshold.py
import sys
import cv2

src_gray = None

threshold_value = 0
threshold_type = 3
max_value = 255
max_type = 4
max_BINARY_value = 255

window_name = 'Threshold Demo'

trackbar_type = 'Type: \n 0: Binary \n 1: Binary Inverted \n' \
                ' 2: Truncate \n 3: To Zero \n 4: To Zero Inverted'
trackbar_value = 'Value'


def main(argv):
    if len(argv) < 1:
        print 'Not enough parameters'
        print 'Usage:\nthreshold.py  < path_to_image >'
        return -1

    # [load_image]
    # Read the image
    src = cv2.imread(argv[1], cv2.IMREAD_COLOR)

    if src is None:
        print 'Invalid input image'
        return -1

    global src_gray
    # Convert it to gray
    src_gray = cv2.cvtColor(src, cv2.COLOR_BGR2GRAY)
    # [load_image]

    # [create_window]
    # Create a window to display results
    cv2.namedWindow(window_name, cv2.WINDOW_AUTOSIZE)
    # [create_window]

    # [create_trackbars]
    # Create Erosion Trackbar
    cv2.createTrackbar(trackbar_type, window_name,
                       threshold_type, max_type, type_update)

    cv2.createTrackbar(trackbar_value, window_name,
                       threshold_value, max_value,
                       value_update)
    # [create_trackbars]

    # Call the function to initialize
    threshold_demo()

    cv2.waitKey(0)
    return 0


def type_update(param):
    global threshold_type
    threshold_type = param
    threshold_demo()


def value_update(param):
    global threshold_value
    threshold_value = param
    threshold_demo()


def threshold_demo():
    """
       0: Binary
       1: Binary Inverted
       2: Threshold Truncated
       3: Threshold to Zero
       4: Threshold to Zero Inverted
    """
    
    ret, dst = cv2.threshold(src_gray, threshold_value,
                             max_BINARY_value, threshold_type)
    cv2.imshow(window_name, dst)


if __name__ == "__main__":
    main(sys.argv)

File: ImgProc/tutorial_threshold/Threshold.cpp
/**
 * @file Threshold.cpp
 * @brief Sample code that shows how to use the diverse threshold options offered by OpenCV
 * @author OpenCV team
 */

#include "opencv2/imgproc/imgproc.hpp"
#include "opencv2/imgcodecs.hpp"
#include "opencv2/highgui/highgui.hpp"
#include <stdlib.h>
#include <stdio.h>

using namespace cv;

/// Global variables

int threshold_value = 0;
int threshold_type = 3;
int const max_value = 255;
int const max_type = 4;
int const max_BINARY_value = 255;

Mat src, src_gray, dst;
const char* window_name = "Threshold Demo";

const char* trackbar_type = "Type: \n 0: Binary \n 1: Binary Inverted \n 2: Truncate \n 3: To Zero \n 4: To Zero Inverted";
const char* trackbar_value = "Value";

/// Function headers
void Threshold_Demo( int, void* );

/**
 * @function main
 */
int main( int, char** argv )
{
  //![load_image]
  /// Load an image
  src = imread( argv[1], 1 );

  /// Convert the image to Gray
  cvtColor( src, src_gray, COLOR_RGB2GRAY );
  //![load_image]

  //![create_window]
  /// Create a window to display results
  namedWindow( window_name, WINDOW_AUTOSIZE );
  //![create_window]

  //![create_trackbars]
  /// Create Trackbar to choose type of Threshold
  createTrackbar( trackbar_type,
                  window_name, &threshold_type,
                  max_type, Threshold_Demo );

  createTrackbar( trackbar_value,
                  window_name, &threshold_value,
                  max_value, Threshold_Demo );
  //![create_trackbars]

  /// Call the function to initialize
  Threshold_Demo( 0, 0 );

  /// Wait until user finishes program
  waitKey(0);

}

//![threshold_demo]
/**
 * @function Threshold_Demo
 */
void Threshold_Demo( int, void* )
{
  /* 0: Binary
     1: Binary Inverted
     2: Threshold Truncated
     3: Threshold to Zero
     4: Threshold to Zero Inverted
   */

  threshold( src_gray, dst, threshold_value, max_BINARY_value,threshold_type );

  imshow( window_name, dst );
}
//![threshold_demo]
